File: quiz/deno/src/statemaschine/states/init-state.ts
import { Select, SelectOption } from '@cliffy/prompt/mod.ts';
import { User } from '@src/user.ts';
import { ExitState } from '@states/exit-state.ts';
import { LoginState } from '@states/login-state.ts';
import { RegisterState } from '@states/register-state.ts';
import { IState } from '@states/state.ts';
import { JoinQuizState } from '@states/quiz/participate/join-quiz-state.ts';

export class InitState implements IState {
	async run(): Promise<IState> {
		enum Action {
			Login = 'login',
			Register = 'register',
			Join = 'join',
			Exit = 'exit',
		}

		const options: (SelectOption<Action>)[] = [];

		if (await User.isAnyUserRegistered()) {
			options.push({ name: 'Login', value: Action.Login });
		}
		options.push({ name: 'Register', value: Action.Register });
		options.push({ name: 'Join Quiz', value: Action.Join });
		options.push({ name: 'Exit', value: Action.Exit });

		const action: string = await Select.prompt({
			message: 'Welcome to THE QUIZ APP',
			options: options,
		});

		switch (action) {
			case Action.Login:
				return new LoginState();
			case Action.Register:
				return new RegisterState();
			case Action.Join:
				return new JoinQuizState();
			case Action.Exit:
				return new ExitState();
		}
		return new InitState();
	}
}

File: quiz/go/statemaschine/states/init-state.go
package states

import (
	"github.com/DoktorLenz/kp-wise23-24/quiz/go/user"
	"github.com/manifoldco/promptui"
)

type InitState struct {
	IState
}

type InitAction string

const (
	InitLogin    InitAction = "Login"
	InitRegister InitAction = "Register"
	InitJoin     InitAction = "Join Quiz"
	InitExit     InitAction = "Exit"
)

func (i *InitState) Run() IState {
	var options []string
	loginAvailable, err := user.IsAnyUserRegistered()
	if err != nil {
		panic(err)
	}
	if loginAvailable {
		options = append(options, string(InitLogin))
	}
	options = append(options, string(InitRegister), string(InitJoin), string(InitExit))

	prompt := promptui.Select{
		Label: "Welcome to THE QUIZ APP",
		Items: options,
	}

	_, result, _ := prompt.Run()

	switch result {
	case string(InitLogin):
		return &LoginState{}
	case string(InitRegister):
		return &RegisterState{}
	case string(InitJoin):
		return &InitState{}
	case string(InitExit):
		return &ExitState{}
	default:
		panic("Unknown action")
	}
}
